File: apps/crawler/src/core/monitors/recruitee.py
"""Recruitee Careers Site API monitor.

Public API: GET https://{slug}.recruitee.com/api/offers
Returns full job data in a single request — no pagination needed.
Also works on custom domains: GET https://{custom-domain}/api/offers
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import TYPE_CHECKING
from urllib.parse import urlparse

import httpx
import structlog

from src.core.enum_normalize import normalize_salary_unit
from src.core.monitors import (
    BoardGoneError,
    DiscoveredJob,
    register,
    slug_guess_allowed,
)
from src.core.monitors._ats_template import ProbeCount, ProbeResult, ats_can_handle
from src.core.monitors.raw import save_json_response
from src.shared.truncation import truncated_rich_result

if TYPE_CHECKING:
    from src.core.monitor import MonitorResult

log = structlog.get_logger()

MAX_JOBS = 50_000

_DOMAIN_RE = re.compile(r"^([\w-]+)\.recruitee\.com$")

_PAGE_PATTERNS = (
    re.compile(r"([\w-]+)\.recruitee\.com"),
    re.compile(r"\b(?:recruiteecdn\.com|window\.recruitee)\b()"),
)

_IGNORE_SLUGS = frozenset({"api", "www", "app", "docs", "help", "support", "status"})

# Recruitee employment-type codes (``fulltime_permanent``,
# ``parttime_fixed_term``, ``freelance``, ``traineeship``, …) pass
# through unchanged — the central
# :func:`src.core.enum_normalize.normalize_employment_type` handles
# them.


def _slug_from_url(board_url: str) -> str | None:
    """Extract the Recruitee company slug from a *.recruitee.com URL."""
    parsed = urlparse(board_url)
    host = (parsed.hostname or "").lower()
    match = _DOMAIN_RE.match(host)
    if match:
        slug = match.group(1)
        if slug not in _IGNORE_SLUGS:
            return slug
    return None


def _api_base_from_url(board_url: str) -> str | None:
    """Derive the API base URL. Returns https://{host} for any Recruitee URL."""
    parsed = urlparse(board_url)
    host = parsed.hostname
    if host:
        scheme = parsed.scheme or "https"
        return f"{scheme}://{host}"
    return None


def _api_url(api_base: str) -> str:
    return f"{api_base}/api/offers"


def _parse_locations(offer: dict) -> list[str] | None:
    """Extract locations from a Recruitee offer."""
    locations: list[str] = []
    seen: set[str] = set()

    # Structured locations array (preferred)
    for loc in offer.get("locations", []):
        city = loc.get("city", "")
        country = loc.get("country", "")
        parts = [p for p in (city, country) if p]
        name = ", ".join(parts)
        if name and name not in seen:
            locations.append(name)
            seen.add(name)

    # Fallback to flat location string
    if not locations:
        flat_loc = offer.get("location")
        if flat_loc and isinstance(flat_loc, str):
            locations.append(flat_loc)

    return locations or None


def _parse_job_location_type(offer: dict) -> str | None:
    """Derive job_location_type from boolean flags."""
    if offer.get("remote"):
        return "remote"
    if offer.get("hybrid"):
        return "hybrid"
    if offer.get("on_site"):
        return "onsite"
    return None


def _parse_salary(offer: dict) -> dict | None:
    """Extract salary from the salary object."""
    salary = offer.get("salary")
    if not salary or not isinstance(salary, dict):
        return None
    sal_min = salary.get("min")
    sal_max = salary.get("max")
    if sal_min is None and sal_max is None:
        return None
    currency = salary.get("currency")
    # Recruitee defaults to ``year`` when period is missing/unknown.
    unit = normalize_salary_unit(salary.get("period")) or "year"
    return {"currency": currency, "min": sal_min, "max": sal_max, "unit": unit}


def _parse_job(offer: dict) -> DiscoveredJob | None:
    """Map a Recruitee offer to a DiscoveredJob."""
    url = offer.get("careers_url")
    if not url:
        return None

    # Combine description + requirements into a single HTML body
    parts: list[str] = []
    desc = offer.get("description")
    if desc:
        parts.append(desc)
    reqs = offer.get("requirements")
    if reqs:
        parts.append(reqs)
    description = "\n".join(parts) if parts else None

    # Employment type — pass through raw upstream code.
    employment_type = offer.get("employment_type_code") or None

    # Metadata
    metadata: dict = {}
    department = offer.get("department")
    if department:
        metadata["department"] = department
    tags = offer.get("tags")
    if tags and isinstance(tags, list):
        metadata["tags"] = tags
    category = offer.get("category_code")
    if category:
        metadata["category"] = category
    offer_id = offer.get("id")
    if offer_id:
        metadata["id"] = offer_id

    return DiscoveredJob(
        url=url,
        title=offer.get("title"),
        description=description,
        locations=_parse_locations(offer),
        employment_type=employment_type,
        job_location_type=_parse_job_location_type(offer),
        date_posted=offer.get("published_at"),
        base_salary=_parse_salary(offer),
        metadata=metadata or None,
    )


async def _probe_api(api_base: str, client: httpx.AsyncClient) -> tuple[bool, int | None]:
    """Probe the Recruitee API. Returns (found, job_count)."""
    try:
        resp = await client.get(_api_url(api_base), follow_redirects=True)
        if resp.status_code != 200:
            return False, None
        data = resp.json()
        offers = data.get("offers")
        if isinstance(offers, list):
            return True, len(offers)
        return False, None
    except Exception:
        log.debug(
            "recruitee.probe_failed",
            probe="api",
            api_base=api_base,
            url=_api_url(api_base),
            exc_info=True,
        )
        return False, None


def _api_base_from_slug(slug: str) -> str | None:
    if not slug:
        return None
    return f"https://{slug}.recruitee.com"


async def _fetch_template_count(
    token: str,
    client: httpx.AsyncClient,
    context: str | None,
) -> ProbeCount | None:
    base = context or _api_base_from_slug(token)
    if base is None:
        return None
    found, count = await _probe_api(base, client)
    return count if found else None


async def _probe_template_slug(
    token: str,
    client: httpx.AsyncClient,
    context: str | None,
) -> ProbeResult:
    _ = context
    base = _api_base_from_slug(token)
    if base is None:
        return False, None
    return await _probe_api(base, client)


async def _probe_page_token(
    token: str,
    client: httpx.AsyncClient,
    context: str | None,
) -> ProbeResult:
    _ = token
    if context is None:
        return False, None
    return await _probe_api(context, client)


def _build_template_result(
    slug: str,
    count: ProbeCount | None,
    api_base: str | None,
) -> dict:
    base = api_base or _api_base_from_slug(slug)
    result: dict = {}
    if slug:
        result["slug"] = slug
    if base is not None:
        result["api_base"] = base
    if count is not None:
        result["jobs"] = count
    return result


async def discover(
    board: dict, client: httpx.AsyncClient, pw=None
) -> list[DiscoveredJob] | MonitorResult:
    """Fetch job listings from the Recruitee public API."""
    metadata = board.get("metadata") or {}
    board_url = board["board_url"]

    # Determine API base: explicit api_base in metadata, or from slug, or from board URL
    api_base = metadata.get("api_base")
    if not api_base:
        slug = metadata.get("slug") or _slug_from_url(board_url)
        api_base = (
            f"https://{slug}.recruitee.com"
            if slug
            else _api_base_from_url(board_url)  # Custom domain
        )

    if not api_base:
        raise ValueError(
            f"Cannot derive Recruitee API base from board URL {board_url!r} "
            "and no slug or api_base in metadata"
        )

    url = _api_url(api_base)
    response = await client.get(url, follow_redirects=True)
    if response.status_code == 404:
        # Recruitee 404s when the company subdomain (slug) has been
        # removed. Surface as a "gone" signal for one-shot disable.
        # See issue #2215.
        raise BoardGoneError(
            f"Recruitee API base {api_base!r} returned 404",
            url=str(response.url),
        )
    response.raise_for_status()

    data = response.json()
    raw_offers = data.get("offers", [])

    jobs: list[DiscoveredJob] = []
    for raw in raw_offers:
        if raw.get("status") != "published":
            continue
        parsed = _parse_job(raw)
        if parsed:
            jobs.append(parsed)

    if len(jobs) > MAX_JOBS:
        log.warning("recruitee.truncated", url=url, total=len(jobs), cap=MAX_JOBS)
        return truncated_rich_result(jobs)

    return jobs


async def can_handle(url: str, client: httpx.AsyncClient | None = None, pw=None) -> dict | None:
    """Detect Recruitee: domain check -> page HTML scan -> slug-based API probe."""
    _ = pw

    def context_from_match(
        match: re.Match[str],
        context: str | None,
    ) -> str | None:
        _ = (match, context)
        return _api_base_from_url(url)

    return await ats_can_handle(
        url,
        client,
        monitor_name="recruitee",
        token_from_url=_slug_from_url,
        page_patterns=_PAGE_PATTERNS,
        ignore_tokens=_IGNORE_SLUGS,
        fetch_job_count=_fetch_template_count,
        api_probe=_probe_template_slug,
        initial_context=None,
        result_builder=_build_template_result,
        context_from_match=context_from_match,
        page_token_probe=_probe_page_token,
        allow_slug_guess=slug_guess_allowed(),
        log_token_field="slug",
    )


async def save_raw(
    artifact_dir: Path,
    board_url: str,
    metadata: dict,
    client: httpx.AsyncClient,
) -> None:
    api_base = metadata.get("api_base")
    if not api_base:
        slug = metadata.get("slug") or _slug_from_url(board_url)
        api_base = f"https://{slug}.recruitee.com" if slug else _api_base_from_url(board_url)
    if not api_base:
        return
    await save_json_response(
        artifact_dir,
        client,
        _api_url(api_base),
        follow_redirects=True,
    )


register("recruitee", discover, cost=10, can_handle=can_handle, rich=True, save_raw=save_raw)

File: apps/crawler/go/recruitee-monitor/recruitee.go
package recruitee

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

const MaxJobs = 50_000

type Job struct {
	URL             string         `json:"url"`
	Title           any            `json:"title"`
	Description     any            `json:"description"`
	Locations       []string       `json:"locations"`
	EmploymentType  any            `json:"employment_type"`
	JobLocationType *string        `json:"job_location_type"`
	DatePosted      any            `json:"date_posted"`
	BaseSalary      map[string]any `json:"base_salary"`
	Metadata        map[string]any `json:"metadata"`
}

type Inventory struct {
	Jobs      []Job `json:"jobs"`
	Truncated bool  `json:"truncated"`
}

func truthy(value any) bool {
	switch v := value.(type) {
	case nil:
		return false
	case bool:
		return v
	case string:
		return v != ""
	case json.Number:
		number, err := v.Float64()
		return err == nil && number != 0
	case []any:
		return len(v) > 0
	case map[string]any:
		return len(v) > 0
	default:
		return true
	}
}

func salaryUnit(value any) string {
	raw, ok := value.(string)
	if !ok {
		return "year"
	}
	key := strings.ToLower(strings.TrimSpace(raw))
	for _, entry := range []struct{ needle, unit string }{
		{"two_weeks", "week"}, {"biweekly", "week"},
		{"hour", "hour"}, {"month", "month"}, {"week", "week"},
		{"year", "year"}, {"annual", "year"}, {"daily", "day"}, {"day", "day"},
	} {
		if strings.Contains(key, entry.needle) {
			return entry.unit
		}
	}
	switch key {
	case "hr", "per-hour-wage":
		return "hour"
	case "mo":
		return "month"
	case "yr":
		return "year"
	}
	return "year"
}

func optionalText(value any, field string) (string, error) {
	if !truthy(value) {
		return "", nil
	}
	text, ok := value.(string)
	if !ok {
		return "", fmt.Errorf("Recruitee %s is not text", field)
	}
	return text, nil
}

func parseLocations(offer map[string]any) ([]string, error) {
	result := []string{}
	seen := map[string]bool{}
	if value, present := offer["locations"]; present {
		locations, ok := value.([]any)
		if !ok {
			return nil, errors.New("Recruitee locations must be a list")
		}
		for _, raw := range locations {
			location, ok := raw.(map[string]any)
			if !ok {
				return nil, errors.New("Recruitee location must be an object")
			}
			city, err := optionalText(location["city"], "location city")
			if err != nil {
				return nil, err
			}
			country, err := optionalText(location["country"], "location country")
			if err != nil {
				return nil, err
			}
			parts := []string{}
			for _, value := range []string{city, country} {
				if value != "" {
					parts = append(parts, value)
				}
			}
			name := strings.Join(parts, ", ")
			if name != "" && !seen[name] {
				result = append(result, name)
				seen[name] = true
			}
		}
	}
	if len(result) == 0 {
		if value, ok := offer["location"].(string); ok && value != "" {
			result = append(result, value)
		}
	}
	if len(result) == 0 {
		return nil, nil
	}
	return result, nil
}

func parseSalary(offer map[string]any) (map[string]any, error) {
	value := offer["salary"]
	if !truthy(value) {
		return nil, nil
	}
	salary, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("Recruitee salary must be an object")
	}
	if salary["min"] == nil && salary["max"] == nil {
		return nil, nil
	}
	return map[string]any{
		"currency": salary["currency"],
		"min":      salary["min"],
		"max":      salary["max"],
		"unit":     salaryUnit(salary["period"]),
	}, nil
}

func parseJob(offer map[string]any) (Job, bool, error) {
	if truthy(offer["careers_url"]) {
		if _, ok := offer["careers_url"].(string); !ok {
			return Job{}, false, errors.New("Recruitee careers_url must be text")
		}
	}
	url, ok := offer["careers_url"].(string)
	if !ok || url == "" {
		return Job{}, false, nil
	}
	locations, err := parseLocations(offer)
	if err != nil {
		return Job{}, false, err
	}
	salary, err := parseSalary(offer)
	if err != nil {
		return Job{}, false, err
	}
	parts := []string{}
	for _, field := range []string{"description", "requirements"} {
		if truthy(offer[field]) {
			value, ok := offer[field].(string)
			if !ok {
				return Job{}, false, fmt.Errorf("Recruitee %s must be text", field)
			}
			parts = append(parts, value)
		}
	}
	var description any
	if len(parts) > 0 {
		description = strings.Join(parts, "\n")
	}
	var employment any
	if truthy(offer["employment_type_code"]) {
		employment = offer["employment_type_code"]
	}
	var locationType *string
	for _, pair := range []struct{ field, value string }{{"remote", "remote"}, {"hybrid", "hybrid"}, {"on_site", "onsite"}} {
		if truthy(offer[pair.field]) {
			value := pair.value
			locationType = &value
			break
		}
	}
	metadata := map[string]any{}
	for _, pair := range []struct{ source, target string }{
		{"department", "department"}, {"category_code", "category"}, {"id", "id"},
	} {
		if truthy(offer[pair.source]) {
			metadata[pair.target] = offer[pair.source]
		}
	}
	if value, ok := offer["tags"].([]any); ok && len(value) > 0 {
		metadata["tags"] = value
	}
	if len(metadata) == 0 {
		metadata = nil
	}
	return Job{
		URL: url, Title: offer["title"], Description: description, Locations: locations,
		EmploymentType: employment, JobLocationType: locationType,
		DatePosted: offer["published_at"], BaseSalary: salary, Metadata: metadata,
	}, true, nil
}

// Parse extracts the same published rich offer fields as the configured
// Recruitee Python monitor. Malformed required input never becomes an empty
// successful inventory.
func Parse(body []byte) (Inventory, error) {
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.UseNumber()
	var root map[string]any
	if err := decoder.Decode(&root); err != nil || root == nil {
		return Inventory{}, errors.New("Recruitee response must be a JSON object")
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return Inventory{}, errors.New("Recruitee response has trailing JSON")
	}
	rawOffers, exists := root["offers"]
	if !exists {
		rawOffers = []any{}
	}
	offers, ok := rawOffers.([]any)
	if !ok {
		return Inventory{}, errors.New("Recruitee offers must be a list")
	}
	result := Inventory{Jobs: make([]Job, 0, len(offers))}
	for index, raw := range offers {
		offer, ok := raw.(map[string]any)
		if !ok {
			return Inventory{}, fmt.Errorf("Recruitee offer %d must be an object", index)
		}
		if offer["status"] != "published" {
			continue
		}
		job, present, err := parseJob(offer)
		if err != nil {
			return Inventory{}, fmt.Errorf("Recruitee offer %d: %w", index, err)
		}
		if present {
			result.Jobs = append(result.Jobs, job)
		}
	}
	result.Truncated = len(result.Jobs) > MaxJobs
	return result, nil
}
